### PR TITLE
[NCL-5611] Initial commit to port to protobuf

### PR DESCRIPTION
**Pull request by thescouser89**
This commit adds classes to properly serialize class:

- MultiValuedMap
- LocalFile

to Protobuf format. `ProtobufSerializer` is the interface annotated with
classes to be considered by the Protobuf Stream processor (Protostream)
to generate proto files for serialization and deserialization.

The Protobuf format is not yet 'activated' in the codebase as we work on
the rest of the classes that need to be serialized via Protobuf.

The spotbug-exclude.xml includes the classes generated by Protostream so
that they are not scanned for checks

### CODE REVIEW

File: core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import org.infinispan.protostream.annotations.ProtoField;
+
+/**
+ * Class wrapping around MultiValuedMap interface (via HashSetValuedHashMap) so that it can be marshalled with Protobuf
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+public class MultiValuedMapProtobufWrapper<K, V> extends HashSetValuedHashMap<K, V> {
+    private static final long serialVersionUID = 20211111L;
```
Comment (dwalluck) [inline core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java:35]: This value doesn't look "right". There's a way to generate the real value as generated by `serialver` in an IDE, or call it manually.

File: core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+     * marshall primitives, enums, and other POJO objects which can be marshalled by Protobuf (as long as they are in
+     * the `AutoProtoSchemaBuilder` list in the ProtobufSerializer interface
+     *
+     * @param <K> Key type
+     * @param <V> Value type
+     *
+     * @see ProtobufSerializer
+     */
+    static class MultiValuedMapProtobufEntry<K, V> {
+
```
Comment (dwalluck) [inline core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java:79]: Small style change here. Remove space here at the beginning of the class and add between the field declarations.

I actually don't understand the style that wants a space there as it makes it look asymmetric to me.

File: core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        private K key;
+        private Collection<V> values;
+
+        // Required by Protostream
+        MultiValuedMapProtobufEntry() {
+        }
```
Comment (dwalluck) [inline core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java:85]: I like to put an empty space here just to make it more clear to me that it's empty.

File: core/src/main/java/org/jboss/pnc/build/finder/protobuf/ProtobufSerializer.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+ */
+package org.jboss.pnc.build.finder.protobuf;
+
+import org.infinispan.protostream.GeneratedSchema;
+import org.infinispan.protostream.annotations.AutoProtoSchemaBuilder;
+import org.jboss.pnc.build.finder.core.LocalFile;
+
+@AutoProtoSchemaBuilder(
+        includeClasses = { LocalFile.class, MultiValuedMapProtobufWrapper.class },
+        schemaFileName = "buildfinder.proto",
```
Comment (dwalluck) [inline core/src/main/java/org/jboss/pnc/build/finder/protobuf/ProtobufSerializer.java:24]: I'd prefer `build-finder.proto` as the name to match the project name.

File: core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        super(map);
+    }
+
+    /**
+     * This method is called for Protobuf -> MultiValuedMapWrapper convertor
+     *
+     * @param entries
+     */
+    @ProtoFactory
+    MultiValuedMapProtobufWrapper(List<MultiValuedMapProtobufEntry<K, V>> entries) {
```
Comment (dwalluck) [inline core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java:52]: Should we use `List` or `Collection`? `Collection` is a bit more flexible and at least here `Map.entrySet()` is a `Set` but it being put into a list. Unfortunately it's not easy to get which type of `Set` it is.

File: core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+     * the `AutoProtoSchemaBuilder` list in the ProtobufSerializer interface
+     *
+     * @param <K> Key type
+     * @param <V> Value type
+     *
+     * @see ProtobufSerializer
+     */
+    static class MultiValuedMapProtobufEntry<K, V> {
+        private K key;
+        private Collection<V> values;
```
Comment (dwalluck) [inline core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java:79]: Blank line here please between fields.

File: core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+                .map(a -> new MultiValuedMapProtobufEntry<K, V>(a.getKey(), a.getValue()))
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * POJO representing the key and list of values of MultiValuedMap for Protobuf.
+     *
+     * The key and values are marshalled as a WrappedMessage. In theory, the WrappedMessage object should know how to
+     * marshall primitives, enums, and other POJO objects which can be marshalled by Protobuf (as long as they are in
+     * the `AutoProtoSchemaBuilder` list in the ProtobufSerializer interface
```
Comment (dwalluck) [inline core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java:70]: I know this is really nitpicking, but can you make sure there is a period at the end of sentences in javadoc? No period after `@`--that is OK.

File: core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+     * Constructor wrapping around the MultiValuedMap
+     *
+     * @param map MultiValuedMap to wrap around
+     */
+    public MultiValuedMapProtobufWrapper(MultiValuedMap<K, V> map) {
+        super(map);
+    }
+
+    /**
+     * This method is called for Protobuf -> MultiValuedMapWrapper convertor
```
Comment (dwalluck) [inline core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java:46]: I forget, does `>` need escaping in Javadoc `&gt;`?

File: core/src/main/java/org/jboss/pnc/build/finder/protobuf/ProtobufSerializer.java (hunk 4 above)
Comment (dwalluck) [reply core/src/main/java/org/jboss/pnc/build/finder/protobuf/ProtobufSerializer.java:24]: You resolved, but I do not see a change.

File: core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java (hunk 8 above)
Comment (thescouser89) [reply core/src/main/java/org/jboss/pnc/build/finder/protobuf/MultiValuedMapProtobufWrapper.java:46]: I'll just remove it

File: core/src/main/java/org/jboss/pnc/build/finder/protobuf/ProtobufSerializer.java (hunk 4 above)
Comment (thescouser89) [reply core/src/main/java/org/jboss/pnc/build/finder/protobuf/ProtobufSerializer.java]: sorry just pushed it again. It got squashed somehow before

**Review by dwalluck — APPROVED** [review]

**Pull request merged**